x/dex/keeper/grpc_query_get_historical_prices.go

```diff
@@ -0,0 +1,120 @@
+package keeper
+
+import (
+	"context"
+	"sort"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/sei-protocol/sei-chain/x/dex/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var ZeroPrice = sdk.ZeroDec()
+
+func (k Keeper) GetHistoricalPrices(goCtx context.Context, req *types.QueryGetHistoricalPricesRequest) (*types.QueryGetHistoricalPricesResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// TODO: optimize by introducing a getter that takes a time filter
+	prices := k.GetAllPrices(ctx, req.ContractAddr, types.Pair{PriceDenom: req.PriceDenom, AssetDenom: req.AssetDenom})
+	currentTimeStamp := uint64(ctx.BlockTime().Unix())
+	beginTimestamp := currentTimeStamp - req.NumOfPeriods*req.PeriodLengthInSeconds
+	// sort descending
+	sort.Slice(prices, func(i, j int) bool {
+		return prices[i].SnapshotTimestampInSeconds > prices[j].SnapshotTimestampInSeconds
+	})
+	validPrices := []*types.Price{}
+	for _, price := range prices {
+		// append early so that we include the latest price before beginTimestamp, since
+		// we need it to set the open price of the first period.
+		if price.SnapshotTimestampInSeconds < currentTimeStamp {
+			validPrices = append(validPrices, price)
+		}
+		if price.SnapshotTimestampInSeconds < beginTimestamp {
+			break
+		}
+	}
+
+	candlesticks := make([]*types.PriceCandlestick, req.NumOfPeriods)
+
+	// set timestamp
+	for i := range candlesticks {
+		candlesticks[i] = &types.PriceCandlestick{}
+		candlesticks[i].EndTimestamp = currentTimeStamp - uint64(i)*req.PeriodLengthInSeconds
+		candlesticks[i].BeginTimestamp = candlesticks[i].EndTimestamp - req.PeriodLengthInSeconds
+	}
+
+	// set open
+	pricePtr := 0
+	for i := range candlesticks {
+		for pricePtr < len(validPrices) && validPrices[pricePtr].SnapshotTimestampInSeconds > candlesticks[i].BeginTimestamp {
```

Comment:
nit: is it better to sort the validPrices in time ascending order so it can reduce some code duplication here? guess the computation is not so important as it's a query endpoint, i'm also good with changing this in a separate PR

Reply:
i intentionally kept open/high/low/close logic separate just to make logic clear for each. If needed I can also consolidate all four computations, but it might be less readable